api/src/main/java/org/pmiops/workbench/ras/RasLinkService.java

```diff
@@ -0,0 +1,147 @@
+ *
+ * <pre>{@code
+ * "sub":"subId",
+ * "name" : "",
+ * "preferred_username":"user1@Login.Gov",
+ * "userId":"user1"
+ * "email":"foo@gmail.com"
+ * }</pre>
+ *
+ * The {@code preferred_username} field should end with "@login.gov" if using that to login. The
+ * {@code email} field is the user used to sign in in login.gov. We can use that as login.gov user
+ * name.
+ *
+ * <p>Step4: Use step3's login.gov username to update AoU database by {@link
+ * UserService#updateRasLinkLoginGovStatus(String)}. Then return it as user profile.
+ *
+ * <p>TODO(yonghao): Fow now we return {@llink ForbiddenException} for all scenarios, determine if
+ * we need to differentiate IAL vs Login.gov scenarios, and give that information to UI.
+ */
+@Service
+public class RasLinkService {
+  private static final Logger log = Logger.getLogger(RasLinkService.class.getName());
+
+  private final UserService userService;
+  private final OpenIdConnectClient rasOidcClient;
+
+  @Autowired
+  public RasLinkService(
+      UserService userService, @Qualifier(RAS_OIDC_CLIENT) OpenIdConnectClient rasOidcClient) {
+    this.userService = userService;
+    this.rasOidcClient = rasOidcClient;
+  }
+
+  /** Links RAS login.gov account with AoU account. */
+  public DbUser linkRasLoginGovAccount(String authCode, String redirectUrl) {
+    JsonNode userInfoResponse = NullNode.getInstance();
+    try {
+      // Oauth dance to get id token and access token.
+      TokenResponse tokenResponse =
+          rasOidcClient.codeExchange(authCode, redirectUrl, RAS_AUTH_CODE_SCOPES);
+
+      // Validate IAL status.
+      String acrClaim =
+          decodedJwt(tokenResponse.get(Id_TOKEN_FIELD_NAME).toString())
+              .getClaim(ACR_CLAIM)
+              .asString();
+      if (!isIal2(acrClaim)) {
+        throw new ForbiddenException(
+            String.format("User does not have IAL2 enabled, acrClaim: %s", acrClaim));
+      }
+      // Fetch user info.
+      userInfoResponse = rasOidcClient.fetchUserInfo(tokenResponse.getAccessToken());
+    } catch (IOException e) {
+      log.log(Level.WARNING, "Failed to link RAS account", e);
+    }
+    return userService.updateRasLinkLoginGovStatus(getLoginGovUsername(userInfoResponse));
+  }
+
+  /** Validates user has IAL2 setup. See class javadoc Step2 for more details. */
+  static boolean isIal2(String acrClaim) {
+    Pattern p = Pattern.compile("ial/\\d", Pattern.CASE_INSENSITIVE);
+    Matcher m = p.matcher(acrClaim);
+    if (m.find()) {
+      return m.group().equals("ial/2");
+    }
+    throw new ServerErrorException(
+        String.format("Invalid acl Claim in OIDC id token for %s", acrClaim));
+  }
+
+  /**
+   * Validates and extracts user's login.gov account from UserInfo response in Json format. See
+   * class javadoc Step3 for more details.
+   */
+  private static String getLoginGovUsername(JsonNode userInfo) {
+    String preferredUsername = userInfo.get(PREFERRED_USERNAME_FIELD_NAME).asText("");
```

Comment:
Mapping this into the expected structure, e.g. a POJO, could perhaps make this more readable

Reply:
This is probably fine, if the responses get more complex, e.g. with passports, we should consider it though. Pulling keys out via string constants can be harder to read in comparison to a class with named properties - specifically there is not an easy place to look to see the full response structure (beyond comments, which can go out of sync).

Reply:
The userInfo is a JWT with some common user information, like email, username, timestamp. Those are simple  enough that don't need pojo. I saw most Oidc library wrapper with those common field plus a generic get(String claimName);  I didn't implement that because the only common thing we need is '`email`'.
Speaking of Passport, that is a complicated Claim(Json), that we  probably want a dedicate pojo, with some method to understand passport. We can add when we need that.